### PR TITLE
Fix deadlocks for distributed training for RichProgressBar

### LINKED ISSUE #10362: `RichProgressBar` deadlocks distributed training
## 🐛 Bug

I have run into many distributed training now that have gone completely frozen after enabling the `RichProgressBar` callback.

It gets stuck in between epochs after it finished one. Sometimes it's after the first one, sometimes after the second one, sometimes after even 9 epochs.

The weird thing is that if I do `Ctrl+C`, the program doesn't interrupt. I have to kill it with SIGKILL (`kill -9`) because SIGTERM doesn't do it. So I don't know the stack trace. I'm also inside a Docker container, so I can't do `strace` (I don't have access to the host computer). Any help here to check the stack trace is welcome.

### To Reproduce

I'm sorry but I can't share my code. All I can say is that I use a machine with 8 GPUs and the `ddp_find_unused_parameters_false` strategy, and that the problem only appears with `RichProgressBar` and that otherwise it doesn't.

I'd be happy to provide more details or try stuff out. Ask me for it! But I can't share my code, I'm sorry.

### Expected behavior

The training to continue successfully.

### Environment

- PyTorch Lightning Version (e.g., 1.3.0): 1.5.0
- PyTorch Version (e.g., 1.8): 1.10.0
- Python version: 3.8.12
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11.3
- GPU models and configuration: 8x A100
- How you installed PyTorch (`conda`, `pip`, source): conda
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #10362 

Fixes were initially part of #9647 with @SeanNaren's efforts
Follow up PR would be to updating the refresh rate and add typing for RichProgress which is missing rn.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -129,13 +129,19 @@ def render(self, task) -> RenderableType:
...
-        def __init__(self, trainer, pl_module):
+        def __init__(self, trainer):
             self._trainer = trainer
-            self._pl_module = pl_module
             self._tasks = {}
             self._current_task_id = 0
+            self.metrics = {}
             super().__init__()
 
+        def update(self, metrics):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/progress/rich_progress.py:139]: Btw, do you know why  mypy is not triggering here? The configuration shows we have rich (and other bars) included. 
cc @carmocca

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by rohitgr7 — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
As discussed offline, we need to validate that this works with users and try to get a test to ensure this works as expected in the future. Hard as current users are unable to share their reproduce!

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 2]
```diff
@@ -129,13 +129,19 @@ def render(self, task) -> RenderableType:
...
+            # called when metrics are ready to be rendered.
+            # this is due to preventing render from causing deadlock issues by requesting metrics
+            # in separate thread.
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py]: Updated!

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 3]
```diff
@@ -129,13 +129,19 @@ def render(self, task) -> RenderableType:
     class MetricsTextColumn(ProgressColumn):
         """A column containing text."""
 
-        def __init__(self, trainer, pl_module):
+        def __init__(self, trainer):
             self._trainer = trainer
-            self._pl_module = pl_module
             self._tasks = {}
             self._current_task_id = 0
+            self.metrics = {}
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py]: Made it protected!

File: pytorch_lightning/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/progress/rich_progress.py:139]: Maybe we need to update it to `"pytorch_lightning.callbacks.progress.*"`?
